### ISSUE TITLE
Benchmark gives NaN for exportable models

### ISSUE THREAD

**Issue by EmmanuelMess** [ultralytics/ultralytics#18904]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Other

### Bug

I am getting this result from benchmark:
```
1300.2s	2103	Benchmarks complete for best.pt on face-detection-dataset.yaml at imgsz=192,320 (536.99s)
1300.2s	2104	                   Format Status❔  Size (MB)  metrics/mAP50-95(B)  Inference time (ms/im)     FPS
1300.2s	2105	0                 PyTorch       ✅        0.3               0.2027                   16.46   60.73
1300.2s	2106	1             TorchScript       ❎        0.7                  NaN                     NaN     NaN
1300.2s	2107	2                    ONNX       ❎        0.5                  NaN                     NaN     NaN
1300.2s	2108	3                OpenVINO       ❎        0.6                  NaN                     NaN     NaN
1300.2s	2109	4                TensorRT       ❌        0.0                  NaN                     NaN     NaN
1300.2s	2110	5                  CoreML       ❎        0.3                  NaN                     NaN     NaN
1300.2s	2111	6   TensorFlow SavedModel       ❎        1.4                  NaN                     NaN     NaN
1300.2s	2112	7     TensorFlow GraphDef       ❎        0.5                  NaN                     NaN     NaN
1300.2s	2113	8         TensorFlow Lite       ❎        0.5                  NaN                     NaN     NaN
1300.2s	2114	9     TensorFlow Edge TPU       ❎        0.3                  NaN                     NaN     NaN
1300.2s	2115	10          TensorFlow.js       ❎        0.5                  NaN                     NaN     NaN
1300.2s	2116	11           PaddlePaddle       ❎        1.0                  NaN                     NaN     NaN
1300.2s	2117	12                    MNN       ❎        0.5                  NaN                     NaN     NaN
1300.2s	2118	13                   NCNN       ✅        0.5               0.0003                    5.97  167.53
1300.2s	2119	14                    IMX       ❌        0.0                  NaN                     NaN     NaN
1300.2s	2120	15                   RKNN       ❌        0.0                  NaN                     NaN     NaN
```

Why do I get so many NaN, especially for TensorFlow Lite, even thought I can export and run the model all right?

Logs:
[logs.log](https://github.com/user-attachments/files/18550557/logs.log)

### Environment

```
Ultralytics 8.3.68 🚀 Python-3.10.14 torch-2.4.0 CUDA:0 (Tesla P100-PCIE-16GB, 16269MiB)
Setup complete ✅ (4 CPUs, 31.4 GB RAM, 6095.9/8062.4 GB disk)

OS                  Linux-6.6.56+-x86_64-with-glibc2.35
Environment         Kaggle
Python              3.10.14
Install             pip
RAM                 31.35 GB
Disk                6095.9/8062.4 GB
CPU                 Intel Xeon 2.00GHz
CPU count           4
GPU                 Tesla P100-PCIE-16GB, 16269MiB
GPU count           1
CUDA                12.3

numpy               ✅ 1.26.4>=1.23.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.7.5>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 11.0.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.4.0>=1.8.0
torch               ✅ 2.4.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.19.0>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 5.9.3
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.12.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0

```

### Minimal Reproducible Example

dataset:
```
%%writefile face-detection-dataset.yaml
# CC0: Public Domain license
# Face-Detection-Dataset dataset by Fares Elmenshawii
# Documentation: https://www.kaggle.com/datasets/fareselmenshawii/face-detection-dataset

# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: /kaggle/input/face-detection-dataset # dataset root dir
train: images/train # train images (relative to 'path')
val: images/val # val images (relative to 'path')
test: # test images (optional)

# Classes
names:
  0: face

# Download script/URL (optional)
download: https://storage.googleapis.com/kaggle-data-sets/3345370/5891144/bundle/archive.zip
```

model:
```
%%writefile yolov6-face.yaml
# Ultralytics YOLO 🚀, AGPL-3.0 license
# YOLOv6 object detection model with P3-P5 outputs. For Usage examples see https://docs.ultralytics.com/models/yolov6

# Parameters
nc: 1 # number of classes
activation: nn.ReLU() # (optional) model default activation function
scales: # model compound scaling constants, i.e. 'model=yolov6n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  p: [0.33, 0.25, 8] # nano is [0.33, 0.25, 1024]

# YOLOv6-3.0s backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]] # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]] # 1-P2/4
  - [-1, 6, Conv, [128, 3, 1]]
  - [-1, 1, Conv, [256, 3, 2]] # 3-P3/8
  - [-1, 12, Conv, [256, 3, 1]]
  - [-1, 1, Conv, [512, 3, 2]] # 5-P4/16
  - [-1, 18, Conv, [512, 3, 1]]
  - [-1, 1, Conv, [1024, 3, 2]] # 7-P5/32
  - [-1, 6, Conv, [1024, 3, 1]]
  - [-1, 1, SPPF, [1024, 5]] # 9

# YOLOv6-3.0s head
head:
  - [-1, 1, Conv, [256, 1, 1]]
  - [-1, 1, nn.ConvTranspose2d, [256, 2, 2, 0]]
  - [[-1, 6], 1, Concat, [1]] # cat backbone P4
  - [-1, 1, Conv, [256, 3, 1]]
  - [-1, 9, Conv, [256, 3, 1]] # 14

  - [-1, 1, Conv, [128, 1, 1]]
  - [-1, 1, nn.ConvTranspose2d, [128, 2, 2, 0]]
  - [[-1, 4], 1, Concat, [1]] # cat backbone P3
  - [-1, 1, Conv, [128, 3, 1]]
  - [-1, 9, Conv, [128, 3, 1]] # 19

  - [[14, 19], 1, Detect, [nc]] # Detect(P3, P4, P5)
```

code:
```
model = YOLO("./yolov6-face.yaml")

r = model.train(data="face-detection-dataset.yaml", epochs=1, imgsz='192,320', single_cls=True, plots=True, batch=500)

from ultralytics.utils.benchmarks import benchmark
benchmark(model=model, data="face-detection-dataset.yaml", imgsz='192,320', device="cpu")
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @EmmanuelMess, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users, where you can find examples for [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage, and where many common questions may already be answered.

If this is a 🐛 Bug Report, thank you for the detailed description! To assist us further in debugging, please ensure you've provided a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/). For benchmarks specifically, this typically includes ensuring:

1. The exact version of `ultralytics` being used. If not already using the latest version (currently **8.3.68**), upgrade with:
   ```bash
   pip install -U ultralytics
   ```
2. The full log file (thank you for attaching it!).
3. The configuration files (i.e., `.yaml` files) and code snippets you've already provided are great, ensuring that someone can replicate your setup.

If this is a custom training ❓ Question about model export, benchmarking, or TensorFlow (TF) inference, please include additional insights like:
- Validation performance for the exported models.
- Details about the environment when running inferences on exported formats, e.g., TensorFlow, ONNX, TensorRT, etc.

### Quick Checks
- Please verify the tools required for benchmarking are installed and match the specified [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml). For example, exporting and testing TensorFlow Lite models requires TensorFlow tooling correctly installed in your environment.
- Cross-check whether your dataset is compatible with multi-format exports. Not all models export equivalently depending on format support.

For more context, you can check the following guides from our Docs:
- [Export Models for Deployment](https://docs.ultralytics.com/modes/export/)

💡 Join the Ultralytics community where it suits you best:
- For real-time discussions and support, join our [Discord](https://discord.com/invite/ultralytics) 🎧.
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/).
- Contribute or discuss insights via [Reddit](https://reddit.com/r/Ultralytics).

## Verified Environments
Benchmark your workflows in one of the verified environments:
- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Docker Image** for consistent dependencies: <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

When the badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing 💚, verifying YOLO operations across platforms.

Thank you for your patience while we look into this! An Ultralytics engineer will review your request and respond shortly. 😊

**Comment by Y-T-G** [comment]
The error is because the benchmark is exporting at a non-square imgsz, while validation does not support non-square imgsz for exported models, so validation fails.

**Comment by EmmanuelMess** [comment]
> The error is because the benchmark is exporting at a non-square imgsz, while validation does not support non-square imgsz for exported models, so validation fails.

Doing `np.int64(np.round(np.sqrt(192*320) / 16)).item() * 16` to get an equivalent square image with the (almost) same size also breaks. 

Doing `np.int64(np.round(np.sqrt(192*320) / 32)).item() * 32` works:

```
Setup complete ✅ (4 CPUs, 31.4 GB RAM, 6100.3/8062.4 GB disk)

Benchmarks complete for ./yolov6-face.yaml on face-detection-dataset.yaml at imgsz=(256, 256) (1169.24s)
                   Format Status❔  Size (MB)  metrics/mAP50-95(B)  Inference time (ms/im)     FPS
0                 PyTorch       ✅        0.3               0.0003                   11.97   83.56
1             TorchScript       ✅        0.7               0.0004                    8.33  120.00
2                    ONNX       ✅        0.5               0.0004                    5.61  178.16
3                OpenVINO       ✅        0.6               0.0003                    4.17  239.87
4                TensorRT       ❌        0.0                  NaN                     NaN     NaN
5                  CoreML       ❎        0.3                  NaN                     NaN     NaN
6   TensorFlow SavedModel       ✅        1.4               0.0003                   11.89   84.09
7     TensorFlow GraphDef       ✅        0.5               0.0003                   10.31   96.97
8         TensorFlow Lite       ✅        0.5               0.0005                    8.32  120.21
9     TensorFlow Edge TPU       ❎        0.4                  NaN                     NaN     NaN
10          TensorFlow.js       ❎        0.5                  NaN                     NaN     NaN
11           PaddlePaddle       ✅        1.0               0.0003                   26.50   37.73
12                    MNN       ✅        0.5               0.0003                    3.53  283.07
13                   NCNN       ✅        0.5               0.0003                    3.75  266.53
14                    IMX       ❌        0.0                  NaN                     NaN     NaN
15                   RKNN       ❌        0.0                  NaN                     NaN     NaN
```

**Comment by EmmanuelMess** [comment]
My recommendations are:
- That you make more clear what ❎ and ❌ mean.
- That you replace the NaN with a "-", because NaN usually means the problem is some division by 0 or equivalent, not that the operation could not be performed
- That you add better warnings, that don't get drowned in gigantic logs (preferably shown at the end) 
- That for broken tests you warn about the reason

**Comment by Y-T-G** [comment]
The log is from the export process and validation. It seems like whatever terminal you used to capture it (or if you dumped it) doesn't properly display the progress bar.

There's a warning when you export with non-square imgsz that validation wouldn't work which also appears when you try to run benchmarks during the export:
```
Export complete (1.7s)
Results saved to /ultralytics
Predict:         yolo predict task=detect model=yolo11n.onnx imgsz=480,640
Validate:        yolo val task=detect model=yolo11n.onnx imgsz=480,640 data=/usr/src/ultralytics/ultralytics/cfg/datasets/coco.yaml WARNING ⚠️ non-PyTorch val requires square images, 'imgsz=[480, 640]' will not work. Use export 'imgsz=640' if val is required.
Visualize:       https://netron.app
```

If you want to just keep warning logs, you can adjust verbosity to only warning:
```python
In [1]: from ultralytics.utils import LOGGER
In [2]: import logging
In [3]: LOGGER.setLevel(logging.WARNING)
```

Each export format has some logs that come from the respective exporter. Ultralytics can't always control the verbosity of those, so you would have to scan the logs to see it.

I guess it would be better to show the error in the table.

**Comment by glenn-jocher** [comment]
Thank you for the clarification. The behavior you're observing is due to how Kaggle handles output streams, which can result in mangled progress bars and logs being redirected to the logs tab. To improve the output experience, you can disable progress bars by setting `tqdm`'s verbosity to minimal using this snippet:

```python
from ultralytics.utils import LOGGER
from tqdm import tqdm
LOGGER.setLevel(logging.WARNING)
tqdm.__init__ = lambda *args, **kwargs: None
```

This should ensure a cleaner output in your Kaggle notebook. Let me know if you need further assistance!

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged via [this PR](https://github.com/ultralytics/ultralytics/pull/18917) – thanks to the awesome contributions of the PR author and contributors! 🎉  

### 🔑 Key Changes in the PR:
1. **Input Size Validation**: Integrated `check_imgsz` to validate square input image sizes during benchmarking, ensuring compatibility and reducing runtime errors.  
2. **Cleaned Outputs**: Replaced `NaN` values in benchmark output tables with a `'-'` placeholder for better readability.  
3. **Improved Logging**: Enhanced status explanations in benchmark results and suppressed verbose output for cleaner benchmarking experiences.  
4. **Usability Enhancements**: Refined the `predict` call to toggle verbosity and improve the smoothness of benchmarking. 📈

To test whether these changes have resolved the issue, you can:  
- Upgrade to the latest GitHub version with:  
  `pip install git+https://github.com/ultralytics/ultralytics.git@main`  
- Or wait for the next official PyPI release.  

Please let us know if this solution addresses the issue as expected. If things persist or new concerns arise, any additional feedback would be warmly welcomed. Thanks again for bringing up this issue—it helps make everything better for everyone! 🙏
